### PR TITLE
feat: Add Map helpers

### PR DESCRIPTION
**Pull request by kkumar-gcc**
RelatedTo: https://github.com/goravel/goravel/issues/82

## 📑 Description 
This pull request introduces a set of map helper methods to enhance map manipulation in Go.

#### Supported Methods
- `Add`: Add an element to a map if it doesn't exist.
- `Exists`: Check if the given key exists in the provided map.
- `Forget`: Remove a given key or keys from the provided map.
- `Get`: Retrieve an element from map.
- `Has`: Check if the given key or keys exist in the provided map.
- `HasAny`: Check if any of the given keys exist in the provided map.
- `Only`: Return the items in the map with the specified keys.
- `Pull`: Return a new map with the specified keys removed.
- `Set`: Set an element to a map using “dot” notation.
- `Where`: filters the items in a map using the given callback.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: support/object/object.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
+package object
```
Comment (hwbrzzl) [inline support/object/object.go:1]: S: Sorry for the change, I checked other packages, and the `maps` name may be better.

File: support/object/object.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	"github.com/gookit/goutil/reflects"
+)
+
+const (
+	Wildcard = "*"
+	PathSep  = "."
+)
+
+// Add an element to a map using “dot” notation if it doesn't exist.
+func Add(obj *map[string]any, k string, v any) error {
```
Comment (hwbrzzl) [inline support/object/object.go:18]: S: Use generic here.

File: support/object/object.go [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+	return Set(obj, k, v)
+}
+
+// Dot flattens a map using dot notation.
+func Dot(obj map[string]any) map[string]any {
+	return maputil.Flatten(obj)
+}
+
+// Exists checks if the given key exists in the provided map (only top level).
+func Exists[K comparable, V any](obj map[K]V, key K) bool {
```
Comment (hwbrzzl) [inline support/object/object.go:32]: 👍

**Review by hwbrzzl — COMMENTED** [review]
Great work, Bro. The functions are basically fine. Just use generic for them if they can.

File: support/object/object.go (hunk 1 above)
Comment (kkumar-gcc) [reply support/object/object.go:1]: No problem, I will change it to maps.

**Comment by hwbrzzl** [comment]
Hi Bro, If this PR is ready for review, please add the `Review Ready` tag.

**Comment by kkumar-gcc** [comment]
As we discussed, I can't implement nested key feature for all cases. So I am not working this feature anymore.

**Comment by hwbrzzl** [comment]
Some of the methods are good, don't you want to merge them?

File: support/maps/maps.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+		if Exists(mp, key) {
+			result[key] = mp[key]
+		}
+	}
+
+	return result
+}
+
+// Pull returns a new map with the specified keys removed.
+func Pull[K comparable, V any](mp map[K]V, key K, def ...any) any {
```
Comment (hwbrzzl) [inline support/maps/maps.go:112]: The action is different with `Arr::pull`:

image

File: support/maps/maps.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+
+	if _, ok := any(key).(string); !ok {
+		return nil
+	}
+
+	value, ok := deleteByPathKeys(mp, mp, strings.Split(any(key).(string), PathSep))
+
+	if !ok && len(def) > 0 {
+		return def[0]
+	}
```
Comment (hwbrzzl) [inline support/maps/maps.go:126]: Remove these lines, we don't need to support dot for now.

File: support/maps/maps.go (hunk 4 above)
Comment (kkumar-gcc) [reply support/maps/maps.go:112]: it's same.

File: support/maps/maps_test.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+			"jane@localhost":  "Jane",
+		},
+	}
+	Forget(aMp, "emails.joe@example.com")
+	assert.Equal(t, map[string]any{
+		"emails": map[string]string{
+			"joe@example.com": "Joe",
+			"jane@localhost":  "Jane",
+		},
+	}, aMp)
```
Comment (hwbrzzl) [inline support/maps/maps_test.go:72]: We don't implement this logic, it can be removed.

File: support/maps/maps_test.go [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+
+	// Test generic type
+	gMp := map[int]string{
+		1: "one",
+		2: "two",
+	}
+	Forget(gMp, 1, 3)
+	assert.Equal(t, map[int]string{
+		2: "two",
+	}, gMp)
```
Comment (hwbrzzl) [inline support/maps/maps_test.go:106]: Great

File: support/maps/maps_test.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+			},
+		},
+		"foo": nil,
+		"bar": map[string]any{
+			"baz": nil,
+		},
+	}
+	assert.True(t, Has(mp, "developers"))
+
+	assert.False(t, Has(mp, "developers", "framework.lang"))
```
Comment (hwbrzzl) [inline support/maps/maps_test.go:180]:
```suggestion
	assert.False(t, Has(mp, "developers", "framework"))
```

File: support/maps/maps_test.go [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+
+	assert.False(t, Has(mp, "framework.dev.name"))
```
Comment (hwbrzzl) [inline support/maps/maps_test.go:182]: We can remove this.

File: support/maps/maps_test.go [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+			"jane@localhost":  "Jane",
+		},
+	}
+	assert.Nil(t, Pull(mp, "emails.joe@example.com"))
+	assert.Equal(t, map[string]any{
+		"emails": map[string]any{
+			"joe@example.com": "Joe",
+			"jane@localhost":  "Jane",
+		},
+	}, mp)
```
Comment (hwbrzzl) [inline support/maps/maps_test.go:273]: This can be removed.

**Review by hwbrzzl — COMMENTED** [review]
Perfect, the test cases are very good. 👍

**Review by hwbrzzl — APPROVED** [review]
Perfect

**Pull request merged**